### ISSUE TITLE
"Variance-covariance matrix of the parameters is not defined" when running clmm

### ISSUE THREAD

**Issue by yanxianl** [runehaubo/ordinal#9]
Dear Rune,

Thank you for developing and maintaining the R package “ordinal”.

I have the following dataset:

    > df
       Fish_ID Tank Diet     Rank
    1      148   A1   FM   Normal
    2      149   A1   FM     Mild
    3      160   A2   FM   Normal
    4      161   A2   FM Moderate
    5      162   A2   FM Moderate
    6      178   A3   FM     Mild
    7      180   A3   FM     Mild
    8      196   A4   FM Moderate
    9      197   A4   FM Moderate
    10     198   A4   FM Moderate
    11     145   B1   IM   Normal
    12     146   B1   IM   Normal
    13     147   B1   IM   Normal
    14     199   B2   IM Moderate
    15     200   B2   IM   Normal
    16     201   B2   IM   Normal
    17     205   B3   IM Moderate
    18     206   B3   IM   Normal
    19     207   B3   IM   Normal
    20     214   B4   IM   Normal
    21     215   B4   IM     Mild
    22     216   B4   IM   Normal
    > str(df)
    'data.frame':	22 obs. of  4 variables:
     $ Fish_ID: chr  "148" "149" "160" "161" ...
     $ Tank   : Factor w/ 8 levels "A1","A2","A3",..: 1 1 2 2 2 3 3 4 4 4 ...
     $ Diet   : Factor w/ 2 levels "FM","IM": 1 1 1 1 1 1 1 1 1 1 ...
     $ Rank   : Ord.factor w/ 3 levels "Normal"<"Mild"<..: 1 2 1 3 3 2 2 3 3 3 ...

I ran the clmm and found a huge condition number of the Hessian. It was mentioned in the tutorial that high condition numbers of the Hessian indicate that the model is ill defined, that the model can be simplified, that possibly some parameters are not identifiable, and that optimization of the model can be difficult. How can I check if my dataset meets the model assumptions of clmm? What can I do in the next steps with a huge Hessian condition number? 

    > mod1 <- clmm(Rank ~ Diet + (1|Tank), data = df)
    > summary(mod1)
    Cumulative Link Mixed Model fitted with the Laplace approximation

    formula: Rank ~ Diet + (1 | Tank)
    data:    df

     link  threshold nobs logLik AIC   niter    max.grad cond.H 
     logit flexible  22   -19.44 46.88 266(288) 2.59e-08 5.9e+06

    Random effects:
     Groups Name        Variance Std.Dev. 
     Tank   (Intercept) 5.83e-09 7.636e-05
    Number of groups:  Tank 8 

    Coefficients:
           Estimate Std. Error z value Pr(>|z|)  
    DietIM  -2.1272     0.9199  -2.312   0.0208 *
    ---
    Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

    Threshold coefficients:
                  Estimate Std. Error z value
    Normal|Mild    -1.1039     0.6627  -1.666
    Mild|Moderate  -0.1227     0.6026  -0.204

Addtionally, when I applied the more accurate approximation to the model, the adaptive Gauss-Hermite quadrature method, I got a warning message in the model summary and many summary statistics are missing. Why is that?

    > mod2 <- clmm(Rank ~ Diet + (1|Tank), data = df, Hess = TRUE, nAGQ = 10)
    > summary(mod2)
    Cumulative Link Mixed Model fitted with the adaptive Gauss-Hermite quadrature approximation with 10 quadrature points 

    formula: Rank ~ Diet + (1 | Tank)
    data:    df

     link  threshold nobs logLik AIC   niter    max.grad cond.H
     logit flexible  22   -19.44 46.88 264(286) 2.84e-08 NaN   

    Random effects:
     Groups Name        Variance  Std.Dev. 
     Tank   (Intercept) 5.431e-09 7.369e-05
    Number of groups:  Tank 8 

    Coefficients:
           Estimate Std. Error z value Pr(>|z|)
    DietIM   -2.127         NA      NA       NA

    Threshold coefficients:
                  Estimate Std. Error z value
    Normal|Mild    -1.1039         NA      NA
    Mild|Moderate  -0.1227         NA      NA
    Warning message:
    In summary.clmm(mod2) :
      Variance-covariance matrix of the parameters is not defined

**Comment by Molly-Scott** [comment]
Hi, did this issue ever get resolved? I am having very similar issues and looking for a solution.

**Comment by acr1408** [comment]
I am also having the same problem. Any further indications how to follow  on this matter. Thanks in advance!

> ordinalModel2 <- clmm(rating ~ sentenceType + (sentenceType|subject) + (sentenceType|item), data=data, threshold = "symmetric")
Using formula(x) is deprecated when x is a character vector of length > 1.
  Consider formula(paste(x, collapse = " ")) instead.
> summary(ordinalModel2)
Variance-covariance matrix of the parameters is not definedCumulative Link Mixed Model fitted with the Laplace approximation

formula: rating ~ sentenceType + (sentenceType | subject) + (sentenceType |      item)
data:    data


Random effects:
 Groups  Name           Variance Std.Dev. Corr                 
 subject (Intercept)     6.3663  2.5231                        
         sentenceTypeCA 11.0408  3.3228   -0.657               
         sentenceTypeCB 28.2928  5.3191   -0.426  0.187        
         sentenceTypeO   1.5614  1.2495   -0.048  0.113 -0.009 
 item    (Intercept)     0.1818  0.4263                        
         sentenceTypeCA  1.8735  1.3688   -0.267               
         sentenceTypeCB  0.1662  0.4076   -0.455  0.980        
         sentenceTypeO   0.1177  0.3431   -0.985  0.095  0.292 
Number of groups:  subject 90,  item 17 

Coefficients:
               Estimate Std. Error z value Pr(>|z|)
sentenceTypeCA  6.15729         NA      NA       NA
sentenceTypeCB -8.65660         NA      NA       NA
sentenceTypeO   0.01205         NA      NA       NA

Threshold coefficients:
          Estimate Std. Error z value
central.1   0.1475         NA      NA
central.2   1.1466         NA      NA
spacing.1   1.8397         NA      NA

**Comment by Yichen-Wu-1997** [comment]
Try to change the random factor into a fixed factor and use clm instead. That's how I solved it.

**Comment by swihart** [comment]
> Try to change the random factor into a fixed factor and use clm instead. That's how I solved it.

I had too many predictors, so I eliminated one and re-ran successfully sticking with `clmm()`.  I acknowledge that may not be an option for some cases.

**Comment by jacobslewis** [comment]
I know this question is closed, but I wanted to make a quick suggestion:

My experience with the "variance-covariance matrix" issue is that usually it is worth reviewing the distribution of either the fixed or random effect ACROSS the level that you want to structure your model at.

So, imagine a really simple bivariate (`y ~ x`) model in which you want to include random intercepts on variable `z`. That model would look like this:

`y ~ x + (1 | z)`

And imagine that z has, say, 10 levels in it. If `x` is singularly bound in a few of those levels, your model will struggle to maximize, and you may arrive at a situation where the model produces a coefficient estimate but not a corresponding standard error (because it can't determine vcov).

This may also be the case when you want to run a bivariate model with both random slopes and intercepts:

`y ~ x + (1 + x | z)`

**So, how do we avoid this?!**

I think the easiest to address this is to run a simple crosstab to see whether `x` is singularly bound across `z`.

`xtabs(~z+x)`

If your table ends up with a significant number of zeroes in certain values of z, then you have a problem.

I hope this helps!
